Sources/StreamChat/AudioPlayer/AssetPropertyLoading.swift

```diff
@@ -0,0 +1,176 @@
+        name = NSExpression(forKeyPath: keyPath).keyPath
+    }
+
+    public var description: String { name }
+}
+
+/// A protocol that describes an object that can be used to load properties from an AVAsset
+public protocol AssetPropertyLoading {
+    /// A method that loads the property of an AVAsset asynchronously and
+    /// returns a result through a completion handler
+    /// - Parameters:
+    ///   - properties: An array containing the properties we would like to load
+    ///   - asset: The asset on which we will try to load the provided properties
+    ///   - completion: The closure to call when we have a final result to report (success or failure)
+    func loadProperties<Asset: AVAsset>(
+        _ properties: [AssetProperty],
+        of asset: Asset,
+        completion: @escaping (Result<Asset, AssetPropertyLoadingCompositeError>) -> Void
+    )
+}
+
+/// A concrete implementation of `AssetPropertyLoading`
+public struct StreamAssetPropertyLoader: AssetPropertyLoading {
+    public func loadProperties<Asset: AVAsset>(
+        _ properties: [AssetProperty],
+        of asset: Asset,
+        completion: @escaping (Result<Asset, AssetPropertyLoadingCompositeError>) -> Void
+    ) {
+        // it's worth noting here that according to the documentation, the completion
+        // handler will be invoked only once, regardless of the number of
+        // properties we are loading.
+        // https://developer.apple.com/documentation/avfoundation/avasynchronouskeyvalueloading/1387321-loadvaluesasynchronously
+        asset.loadValuesAsynchronously(forKeys: properties.map(\.name)) {
+            handlePropertiesLoadingResult(
+                properties,
+                of: asset,
+                completion: completion
+            )
+        }
+    }
+
+    /// A private method that handles the result of loading the properties of an AVURLAsset and returns a
+    /// result through a completion handler
+    private func handlePropertiesLoadingResult<Asset: AVAsset>(
+        _ properties: [AssetProperty],
+        of asset: Asset,
+        completion: @escaping (Result<Asset, AssetPropertyLoadingCompositeError>) -> Void
+    ) {
+        var failedProperties: [AssetPropertyLoadingFailedError] = []
+        var cancelledProperties: [AssetPropertyLoadingCancelledError] = []
+
+        properties.forEach { property in
+            var error: NSError?
+            let statusOfValue = asset.statusOfValue(
+                forKey: property.name,
+                error: &error
+            )
+
+            /// Handle the status of the loaded property and call the appropriate completion handler
+            switch statusOfValue {
+            case .loading:
+                /// Do nothing if the property is still loading
+                break
+            case .loaded:
+                /// If the property has been loaded, apply the onSuccessTransformer to the asset and call
```

Comment:
The comment here doesn't really match the code. Is it only "do nothing" in this case or sth is forgotten? If the first, we could also put it together with `loading` or maybe explain that an action would be executed after the switch.

Reply:
Good catch. This is now updated